### PR TITLE
Current Interpolation: During add to EMF

### PR DESCRIPTION
**Pull request by ax3l**
This commit adds the general infrastructure to call functors before adding `FieldJ` to the electro-magnetic fields (in `Yee`: before manipulating the `E` field).

The previous method `None` without interpolation is added as a default in `fieldSolver.param`. Additionally, a simple second order Binomial filter for 2D and 3D is added to test the new functionality.

Run time tests:
- [x] KHI 3D e+/e- `dev`vs. with/without Binomial filter (with ZigZag & TSC)

### CODE REVIEW

File: src/picongpu/include/fields/currentInterpolation/Binomial/Binomial.hpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    HDINLINE void operator()(DataBoxE fieldE,
+                             DataBoxB,
+                             DataBoxJ fieldJ )
+    {
+        const DataSpace<MappingDesc::Dim> self;
+        typedef typename DataBoxJ::ValueType TypeJ;
+
+        /* 1 2 1 weighting for "left"(1x) "center"(2x) "right"(1x),
+         * see Pascal's triangle level N=2 */
+        const TypeJ dirX1 = fieldJ(DataSpace<DIM3>(-1, 0, 0)) + fieldJ(self);
```
Comment (ax3l) [inline src/picongpu/include/fields/currentInterpolation/Binomial/Binomial.hpp]: @psychocoderHPC: for loop over `simDim` -> avoids second impl

**Comment by ax3l** [comment]
@psychocoderHPC event system updates added.

File: src/picongpu/include/fields/FieldJ.tpp [hunk 2]
```diff
@@ -101,10 +115,34 @@ fieldJ( cellDescription.getGridLayout( ) ), fieldE( NULL )
...
+            for ( uint32_t d = 0; d < simDim; ++d )
+                guardingCells[d] = ( relativMask[d] == -1 ? originRecvGuard[d] : endRecvGuard[d] );
+            fieldJrecv->addExchange( GUARD, i, guardingCells, FIELD_JRECV );
+        }
+    }
 }
 
 FieldJ::~FieldJ( )
 {
+    __delete(fieldJrecv);
```
Comment (ax3l) [inline src/picongpu/include/fields/FieldJ.tpp]: Should be ok for [NULL](https://github.com/ComputationalRadiationPhysics/picongpu/blob/e58cf8271c7e057a128bce172b9b987ba6bc44ee/src/libPMacc/include/types.h#L173) values when unused.

Also see: #532

File: src/picongpu/include/simulationControl/MySimulation.hpp [hunk 3]
```diff
@@ -405,13 +415,33 @@ class MySimulation : public SimulationHelper<simDim>
...
-            fieldJ->addCurrentToE<CORE > ();
 
-            __setTransactionEvent(eRecvCurrent);
-            fieldJ->addCurrentToE<BORDER > ();
+            const DataSpace<simDim> currentRecvLower( GetMargin<fieldSolver::CurrentInterpolation>::LowerMargin( ).toRT( ) );
+            const DataSpace<simDim> currentRecvUpper( GetMargin<fieldSolver::CurrentInterpolation>::UpperMargin( ).toRT( ) );
+
+            /* without interpolation, we do not need to access the FieldJ GUARD
+             * and can therefor overlap communication of GUARD->(ADD)BORDER & computation of CORE */
+            if( currentRecvLower != DataSpace<simDim>::create(0) ||
```
Comment (ax3l) [inline src/picongpu/include/simulationControl/MySimulation.hpp]: wrong! inverse!

File: src/picongpu/include/fields/FieldJ.tpp (hunk 2 above)
Comment (psychocoderHPC) [inline src/picongpu/include/fields/FieldJ.tpp]: Yes `__delete` checks for `NULL`

Comment (ax3l) [inline src/picongpu/include/fields/FieldJ.tpp]: even [delete()](http://en.cppreference.com/w/cpp/memory/new/operator_delete) itself checks for null (`nullptr` and `NULL` are well-defined input for the std `delete`), that is why `__delete()` should not double-check it (see #532)

**Comment by ax3l** [comment]
*Working on test* ...

Yee solver, KHI default but with e+/e-:
  `dev` vs `pull` with `None` vs `pull` with `Binomial`

`-d 2 8 1 -g 192 512 16 -s 2000` (hdf5 every 250)
full simulation time: `...` `564` `561` sec

...

**Comment by ax3l** [comment]
the main difference might be due to the seed, since the species ID is used in it (dev to None).

The binomial image looks "smoother" with less edgy features in the B-fields, which I would expect and is quite physical correct.

**Comment by psychocoderHPC** [comment]
Is it correct that the CFL is near 1.0?  
You wrote: `Courant c*dt <= 1.02768 ? 1`

**Comment by psychocoderHPC** [comment]
Only a note: In general a forward declaration like `currentInterpolation/None/None.def` and `currentInterpolation/Binomial/Binomial.def` is not needed because the solver itself is free from dependencies.
It is also possible to write and use only `*.hpp` because everything is given in by templates.

**Comment by ax3l** [comment]
same as before, no hdf5, png every 100 (instead of 10):

else still: binning every 100, energy sum of fields & particles every 10.

full simulation time: `341` `341` `346` sec (order: `dev`, `none`, `bin.`)

khi_currentfilter_ene
t=1000:
khi_currentfilter

**Comment by ax3l** [comment]
ok that looks valid (I now keep the seed change as we planned). the ene is the same for `dev` and the new `None`.

the only small difference is a tiny bit different scaling in the png preview, that I don't see an immediate reason for.

**Comment by psychocoderHPC** [comment]
:+1:

File: src/picongpu/include/fields/currentInterpolation/Binomial/Binomial.hpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        const DataSpace<dim> self;
+        typedef typename DataBoxJ::ValueType TypeJ;
+
+        /* 1 2 1 weighting for "left"(1x) "center"(2x) "right"(1x),
+         * see Pascal's triangle level N=2 */
+        TypeJ dirSum( TypeJ::create(0.0) );
+        for( uint32_t d = 0; d < dim; ++d )
+        {
+            DataSpace<dim> dw;
```
Comment (psychocoderHPC) [inline src/picongpu/include/fields/currentInterpolation/Binomial/Binomial.hpp]: If you need const `DataSpace<>` this is the way. (only a not, no need to put it in)
```C++   
 typedef typename PMacc::math::CT::make_Vector<
        simDim,
        bmpl::integral_c<int,0>
    >::type ZeroVector;

    typedef typename PMacc::math::CT::Assign<
        ZeroVector,
        bmpl::integral_c<int,0>,
        bmpl::integral_c<int,-1>
    >::type Minus;

    typedef typename PMacc::math::CT::Assign<
        ZeroVector,
        bmpl::integral_c<int,0>,
        bmpl::integral_c<int,1>
    >::type Plus;

        const PMACC_AUTO(minus,Minus::toRT());
        const PMACC_AUTO(plus,Plus::toRT());
```

**Pull request merged**